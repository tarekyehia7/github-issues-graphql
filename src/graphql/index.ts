import { ApolloClient, createHttpLink, InMemoryCache } from '@apollo/client';
import { setContext } from '@apollo/client/link/context';

const httpLink = createHttpLink({
	uri: process.env.REACT_APP_API_URL,
});

const authLink = setContext((_, { headers }) => {
	const token = process.env.REACT_APP_GITHUB_AUTH_TOKEN;

	return {
		headers: {
			...headers,
			authorization: token ? `Bearer ${token}` : '',
		},
	};
});

export const apolloClient = new ApolloClient({
	link: authLink.concat(httpLink),
	cache: new InMemoryCache(),
	connectToDevTools: true,
});
