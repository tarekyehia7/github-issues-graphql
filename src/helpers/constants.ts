export const constants = {
    repository: 'facebook',
    projectName: 'react',
	searchTerm: 'title,body',
    issuesPerPage: 10,
    githubUrl: 'https://github.com'
};