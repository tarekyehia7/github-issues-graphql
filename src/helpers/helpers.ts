import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';

export const getPagesNumber = (totalIssuesCount: number, issuesPerPage: number): number => {
	const totalPages = Math.ceil(totalIssuesCount / issuesPerPage);
	return !totalPages ? 1 : totalPages;
};

export const formatDate = (date: string) => {
	try {
		const myDate = new Date(date);
		const myFormattedDate = formatDistanceToNowStrict(myDate, { addSuffix: true });
		return myFormattedDate;
	} catch (e) {
		console.error(e);
		return '';
	}
};
