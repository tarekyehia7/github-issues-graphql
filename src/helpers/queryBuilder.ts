import { StatusEnum } from "../pages/Issues/Issues";

interface Query {
    repo: string;
    input: string;
    is: string;
    in: string;
    status: StatusEnum;
}

export const buildQuery = (query: Query) => {
    return Object.entries(query)
        .reduce((str, [p, val]) => `${str}${p}:${val} `, '')
        .replace('status:', '')
        .replace('input:', '')
        .trim();
};