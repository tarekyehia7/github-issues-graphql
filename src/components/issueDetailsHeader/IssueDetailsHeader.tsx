import formatDistanceToNowStrict from 'date-fns/formatDistanceToNowStrict';
import React from 'react';
import styled from 'styled-components';
import { IssueState } from '../../graphql/generatedTypes/graphql';
import { OpenIssueIcon } from '../../icons/OpenIssueIcon';
import { SkipIcon } from '../../icons/SkipIcon';

const TitleContainer = styled.div`
    font-weight: 400;
    padding-bottom: 3rem;
    border-bottom: 1px solid #d0d7de;
`;

const BigTitle = styled.h1`
`;

const Description = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;

const Status = styled.span<{ isOpenState: boolean }>`
    display: flex;
    align-items: center;
    justify-content: space-between;
    width: ${({ isOpenState }) => isOpenState ? '4rem' : '5rem'};
    border: 1px solid transparent;
    color: #ffffff;
    padding: 5px 12px;
    font-size: 14px;
    font-weight: 500;
    line-height: 20px;
    text-align: center;
    white-space: nowrap;
    border-radius: 2em;
    background-color: ${({ isOpenState }) => isOpenState ? '#fa4549' : '#6e7781'};
    svg {
        fill: #ffffff
    }
`;

type IssueHeaderProps = {
    title: string;
    issueId: string;
    state: IssueState;
    authorName: string;
    createdAt: string;
};

export const IssueDetailsHeader = ({
    title,
    issueId,
    state,
    authorName,
    createdAt
}: IssueHeaderProps) => (
    <TitleContainer>
        <BigTitle>{title} #{issueId}</BigTitle>
        <Description>
            <Status isOpenState={state === IssueState.Open}>
                {state === IssueState.Open ?
                    <OpenIssueIcon />
                    :
                    <SkipIcon />
                }
                {state} 
            </Status>
            {authorName} opened this issue {formatDistanceToNowStrict(new Date(createdAt))} ago
        </Description>
    </TitleContainer>
);