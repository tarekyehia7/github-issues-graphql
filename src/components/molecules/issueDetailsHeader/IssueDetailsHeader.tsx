import React from 'react';

import { Description, TitleContainer } from './IssueDetailsHeader.styled';
import { IssueState } from '../../../graphql/generatedTypes/graphql';
import { formatDate } from '../../../helpers/helpers';
import { Title, TitleSizeTypes } from '../../atoms/Title/Title';
import { Status, StatusEnum } from '../../atoms/status/Status';

export type IssueHeaderProps = {
	title: string;
	issueId: string;
	state: IssueState;
	authorName: string;
	createdAt: string;
};

export const IssueDetailsHeader = ({
	title,
	issueId,
	state,
	authorName,
	createdAt,
}: IssueHeaderProps) => {
	const status = state === IssueState.Open ? StatusEnum.Open : StatusEnum.Closed;

	return (
		<TitleContainer>
			<Title titleSize={TitleSizeTypes.Large}>
				<span>{title}</span> #{issueId}
			</Title>
			<Description>
				<Status status={status} />
				{authorName} opened this issue {formatDate(createdAt)}
			</Description>
		</TitleContainer>
	);
};
