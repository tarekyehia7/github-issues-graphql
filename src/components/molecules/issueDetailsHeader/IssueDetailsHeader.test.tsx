import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import 'jest-styled-components';

import { IssueDetailsHeader, IssueHeaderProps } from './IssueDetailsHeader';
import { PageWithTheme } from '../../../helpers/testing/helpers';
import { IssueState } from '../../../graphql/generatedTypes/graphql';
import { formatDate } from '../../../helpers/helpers';

const props: IssueHeaderProps = {
	state: IssueState.Open,
	title: 'Bug: <img/> renders twice in firefox with react 18',
	issueId: '25276',
	createdAt: '2022-09-09T13:25:28Z',
	authorName: 'tyehia',
};

const renderPage = () => {
	const { container, getByText } = render(
		<PageWithTheme>
			<IssueDetailsHeader {...props} />
		</PageWithTheme>,
	);

	return {
		container,
		getByText,
	};
};

describe('<IssueDetailsHeader />', () => {
	it('Should match snapshot', () => {
		const { container } = renderPage();
		expect(container).toMatchSnapshot();
	});

	it('Should have correct Title', () => {
		const { getByText } = renderPage();

		expect(getByText(props.title)).toBeInTheDocument();
		expect(getByText(`#${props.issueId}`)).toBeInTheDocument();
	});

	it('Should have correct description', () => {
		const { getByText } = renderPage();

		expect(
			getByText(`${props.authorName} opened this issue ${formatDate(props.createdAt)}`),
		).toBeInTheDocument();
	});
});
