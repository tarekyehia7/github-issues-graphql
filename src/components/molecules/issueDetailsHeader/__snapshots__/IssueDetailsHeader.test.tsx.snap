// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IssueDetailsHeader /> Should match snapshot 1`] = `
.c0 {
  font-weight: 400;
  padding-bottom: 3rem;
  border-bottom: 1px solid #d0d7de;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 4rem;
  border: 1px solid transparent;
  color: #ffffff;
  padding: 5px 12px;
  font-size: 14px;
  font-weight: 500;
  line-height: 20px;
  text-align: center;
  white-space: nowrap;
  border-radius: 2em;
  background-color: #fa4549;
  margin-right: 0.7rem;
  width: 3rem;
}

.c2 svg {
  fill: #ffffff;
}

<div>
  <div
    class="c0"
  >
    <h1
      class=""
    >
      <span>
        Bug: &lt;img/&gt; renders twice in firefox with react 18
      </span>
       #
      25276
    </h1>
    <div
      class="c1"
    >
      <span
        class="c2"
      >
        <svg
          aria-hidden="true"
          data-testid="open-issue-icon"
          data-view-component="true"
          height="16"
          version="1.1"
          viewBox="0 0 16 16"
          width="16"
        >
          <path
            d="M8 9.5a1.5 1.5 0 100-3 1.5 1.5 0 000 3z"
          />
          <path
            d="M8 0a8 8 0 100 16A8 8 0 008 0zM1.5 8a6.5 6.5 0 1113 0 6.5 6.5 0 01-13 0z"
          />
        </svg>
        open
      </span>
      tyehia
       opened this issue 
      9 days
       ago
    </div>
  </div>
</div>
`;
